### PR TITLE
added tests, modified implements of OperationHandler

### PR DESCRIPTION
**Pull request by Daniel1004K**

### CODE REVIEW

File: src/test/java/core/basesyntax/service/CsvFileReaderTest.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        expected.add("b,banana,20");
+        expected.add("b,apple,100");
+        expected.add("s,banana,100");
+        expected.add("p,banana,13");
+        expected.add("r,apple,10");
+        expected.add("p,apple,20");
+        expected.add("p,banana,5");
+        expected.add("s,banana,50");
+        List<String> actual = fileReader.readFromFile(PATH_FILE_INTO);
+        assertArrayEquals(expected.toArray(), actual.toArray());
```
Comment (Tamara-Paskalova) [inline src/test/java/core/basesyntax/service/CsvFileReaderTest.java:32]: You can compare two Lists with assertEquals method. It do the same and you don't have to cast ArrayList into Array

File: src/test/java/core/basesyntax/service/FileWriterTest.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        fileWriter = new FileWriter();
+    }
+
+    @Test
+    public void fileWriter_successfulWrite_ok() {
+        fileWriter.writeToFile(REPORT_VALID, PATH_FILE_REPORT);
+        try {
+            String actual = String
+                    .join(System.lineSeparator(), Files.readAllLines(Path.of(PATH_FILE_REPORT)));
+            assertEquals(REPORT_VALID, actual);
```
Comment (Tamara-Paskalova) [inline src/test/java/core/basesyntax/service/FileWriterTest.java:33]: Logic of assertEquals do not relate to ligic of try-catch block. Let's move it out of here

File: src/test/java/core/basesyntax/service/FileWriterTest.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
```
Comment (Tamara-Paskalova) [inline src/test/java/core/basesyntax/service/FileWriterTest.java:36]: Bad practice

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]
test

**Review by Tamara-Paskalova — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/Main.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import core.basesyntax.strategy.AdditionHandler;
+import core.basesyntax.strategy.BalanceHandler;
+import core.basesyntax.strategy.OperationHandler;
+import core.basesyntax.strategy.PurchaseHandler;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class Main {
+
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Main.java:22]: redundant empty line

File: src/main/java/core/basesyntax/dbtest/Storage.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package core.basesyntax.dbtest;
+
+import core.basesyntax.model.Fruit;
+import java.util.HashMap;
+import java.util.Map;
+
+public class Storage {
+
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/dbtest/Storage.java:8]: same

File: src/main/java/core/basesyntax/dto/FruitDto.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
+package core.basesyntax.dto;
+
+import java.util.Objects;
+
+public class FruitDto {
+
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/dto/FruitDto.java:6]: same

File: src/main/java/core/basesyntax/model/Fruit.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
+package core.basesyntax.model;
+
+import java.math.BigDecimal;
+import java.util.Objects;
+
+public class Fruit {
+
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/model/Fruit.java:7]: same

File: src/main/java/core/basesyntax/service/FileReader.java [hunk 8]
```diff
@@ -0,0 +1,8 @@
+package core.basesyntax.service;
+
+import java.util.List;
+
+public interface FileReader {
+
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/service/FileReader.java:6]: same

File: src/main/java/core/basesyntax/service/FileWriter.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+public class FileWriter implements Writer {
+    private static final String PATTERN = "(\\w+,\\w+\\r*\\n*)(\\w+,\\d+\\r*\\n*)*";
+
+    @Override
+    public boolean writeToFile(String report, String filePath) {
+
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/service/FileWriter.java:13]: same(
please check all your classes, both source and test

we don't need empty lines after method signatures or class declarations

File: src/test/java/core/basesyntax/strategy/PurchaseHandlerTest.java [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+import org.junit.Before;
+import org.junit.Test;
+
+public class PurchaseHandlerTest {
+    public static final String PATH_INPUT_FILE =
+            "src/test/resources/purchasetest/storagePurchaseTest.csv";
+    public static final String PATH_OUTPUT_FILE =
+            "src/test/resources/purchasetest/resultPurchaseTest.csv";
+    public static final String PATH_FILE_RESULT =
+            "src/test/resources/purchasetest/reportPurchaseTest.csv";
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/strategy/PurchaseHandlerTest.java:18]: please remove fields that you do not use

File: src/test/java/core/basesyntax/strategy/PurchaseHandlerTest.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+    @Before
+    public void setStorageOperation() {
+        StorageTest.storage.clear();
+    }
+
+    @Test
+    public void purchaseHandler_checkGettingObject_notOk() {
+        FruitDto fruitDto = new FruitDto("b", "banana", 100);
+        OperationHandler expected = new PurchaseHandler(StorageTest.storage);
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/strategy/PurchaseHandlerTest.java:28]: let's make a field and initialize it in `@BeforeClass` method

File: src/test/java/core/basesyntax/strategy/BalanceHandlerTest.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Before
+    public void setUp() {
+        StorageTest.storage.clear();
+    }
+
+    @Test
+    public void balanceHandler_checkGettingObject_notOk() {
+        FruitDto fruitDto = new FruitDto("s", "banana", 100);
+        OperationHandler expected = new BalanceHandler(StorageTest.storage);
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/strategy/BalanceHandlerTest.java:29]: same for all handlers

File: src/test/java/core/basesyntax/service/FruitReportServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+import static org.junit.Assert.assertEquals;
+
+import core.basesyntax.dbtest.StorageTest;
+import core.basesyntax.model.Fruit;
+import org.junit.Before;
+import org.junit.Test;
+
+public class FruitReportServiceTest {
+    private static FruitReportService fruitReportService;
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/service/FruitReportServiceTest.java:11]: let's use abstraction

File: src/test/java/core/basesyntax/service/FruitReportServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+    @Before
+    public void setFruitReportService() {
+        StorageTest.storage.clear();
+        fruitReportService = new FruitReportService(StorageTest.storage);
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/service/FruitReportServiceTest.java:16]: move initialization to `@BeforeClass`

File: src/test/java/core/basesyntax/service/FileWriterTest.java [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+            + "apple90\n"
+            + "banana152";
+    private static final String REPORT_VALID = "fruit,quantity" + System.lineSeparator()
+            + "apple,90" + System.lineSeparator()
+            + "banana,152";
+    private FileWriter fileWriter;
+
+    @Before
+    public void setFileWriter() {
+        fileWriter = new FileWriter();
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/service/FileWriterTest.java:25]: same

File: src/test/java/core/basesyntax/service/FileWriterTest.java [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+        StorageTest.storage.clear();
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/service/FileWriterTest.java:26]: we are not communicating with storage in this test class

File: src/test/java/core/basesyntax/service/CsvFileReaderTest.java [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import org.junit.Before;
+import org.junit.Test;
+
+public class CsvFileReaderTest {
+    private static final String PATH_FILE_INTO = "src/main/resources/storage.csv";
+    private CsvFileReader fileReader;
+
+    @Before
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/service/CsvFileReaderTest.java:15]:
```suggestion
    @BeforeClass
```

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/strategy/PurchaseHandlerTest.java (hunk 11 above)
Comment (KseniiaMakarova) [reply src/test/java/core/basesyntax/strategy/PurchaseHandlerTest.java:21]: > let's make a field and initialize it in @BeforeClass method

what about this comment?

File: src/test/java/core/basesyntax/strategy/BalanceHandlerTest.java (hunk 12 above)
Comment (KseniiaMakarova) [reply src/test/java/core/basesyntax/strategy/BalanceHandlerTest.java:22]: > same for all handlers

 and this?

File: src/test/java/core/basesyntax/strategy/AdditionHandlerTest.java [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+public class AdditionHandlerTest {
+    @Before
+    public void clearStorage() {
+        StorageTest.storage.clear();
+    }
+
+    @Test
+    public void additionHandler_checkGettingObject_notOk() {
+        FruitDto fruitDto = new FruitDto("b", "banana", 100);
+        OperationHandler expected = new AdditionHandler(StorageTest.storage);
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/strategy/AdditionHandlerTest.java:21]: here as well

File: src/test/java/core/basesyntax/service/FileWriterTest.java [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+public class FileWriterTest {
+    private static final String PATH_FILE_REPORT =
+            "src/test/resources/csv_file_writer_test/storage.csv";
+    private static final String REPORT_IS_NOT_VALID = "fruit_quantity" + System.lineSeparator()
+            + "apple90\n"
+            + "banana152";
+    private static final String REPORT_VALID = "fruit,quantity" + System.lineSeparator()
+            + "apple,90" + System.lineSeparator()
+            + "banana,152";
+    private static FileWriter fileWriter;
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/service/FileWriterTest.java:20]: use abstraction please

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/strategy/AdditionHandlerTest.java [hunk 20]
```diff
@@ -6,49 +6,48 @@
...
 public class AdditionHandlerTest {
-    @Before
-    public void clearStorage() {
+    private static OperationHandler additionHandler;
+
+    @BeforeClass
+    public static void clearStorage() {
+        additionHandler = new AdditionHandler(StorageTest.storage);
         StorageTest.storage.clear();
     }
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/strategy/AdditionHandlerTest.java:19]: now in all handlers the storage is cleared once before all methods - we should do this before/after each test

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request closed without merging**